Turn on the `Copyable` as an inferred generic constraint by default

kavon opened the pull request:
This is the main thing required before we can make noncopyable / move-only types available by default (without a flag).

resolves rdar://104898230

kavon committed:
revise how the `Copyable` constraint is added

The _Copyable constraint was implemented as a marker protocol.
That protocol is part of the KnownProtocol's in the compiler.
When `ASTContext::getProtocol(KnownProtocolKind kind)` tries
to find the ProtocolDecl for Copyable, it will look in the
stdlib module (i.e., Swift module), which is where I initially
planned to put it.

That created problems initially when some regression tests
use `-parse-stdlib` failed to do that protocol lookup, which is
essential for adding the constraint (given the current implementation).

That led to believe we need to pull Copyable out of the stdlib, but that's
wrong. In fact, when building the Swift module itself, we do `-parse-stdlib`
but we also include `-module-name Swift`. This causes the _Copyable protocol
defined in the Stdlib to be correctly discovered while building the stdlib
itself (see the test case in this commit). So, the only downside of
having the Copyable protocol in the Stdlib is that `-parse-stdlib` tests
in the compiler can't use move-only types correctly, as they'll be
allowed in generic contexts. No real program would build like this.

Until I have time to do a further refactoring, this is an acceptable trade-off.

fixes rdar://104898230

kavon committed:
[nfc] remove outdated FIXME's

ktoso reviewed (approved):
Huh, sweet, thanks for the fix @kavon !

kavon committed:
add a Copyable conformance for `PackExpansionType`

slavapestov commented on lib/AST/ASTContext.cpp:
```diff
@@ -1103,8 +1103,29 @@ ProtocolDecl *ASTContext::getProtocol(KnownProtocolKind kind) const {
+  // The Copyable protocol is not defined in any module. In particular, we can't
+  // define it in the stdlib because it's required to exist in order to
+  // typecheck all parts of the stdlib, so you would have to encounter the decl
+  // first. So we generate it on-demand and pretend it is in the stdlib.
```
I'm not sure why it's any different from, eg, Sendable. What doesn't work if its declared in source?

slavapestov replied:
Ok, I see. The -parse-stdlib issue isn't a bootstrapping problem where you "haven't encountered the decl first", it's literally that test cases that use -parse-stdlib don't load the real stdlib, so they should define the dummy protocol. I still think that's cleaner than synthesizing it like this.

kavon replied:
Hm okay. I thought maybe `-parse-stdlib` was a more fundamentally required functionality of the compiler, but what you said makes it seem like it's mostly internally used for testing. So for now I'll go with the "don't add constraint if it's missing" since that's cleaner.

kavon replied:
actually we do build parts of the stdlib with `-parse-stdlib` so I'm going to keep this synthesizing hack.

slavapestov commented on lib/Sema/TypeCheckType.cpp:
```diff
@@ -4128,6 +4126,18 @@ TypeResolver::resolveDeclRefTypeRepr(DeclRefTypeRepr *repr,
   }
 
+  // Raise an error if anyone explicitly refers to the Copyable type, anywhere.
+  // We're not ready yet for anyone to express these kinds of constraints.
```
Maybe just put an UnavailableAttr on it when you synthesize it?

kavon replied:
good idea. thanks! this works well.

kavon commented:
@swift-ci please test

slavapestov commented on lib/AST/Module.cpp:
```diff
@@ -1688,9 +1688,12 @@ LookupConformanceInModuleRequest::evaluate(
+        // this stage, we don't even want Copyable to appear in swiftinterface
+        // files, which will happen for a marker protocol that's registered
+        // in a nominal type's conformance table. We can reconsider this
+        // decision later once there's a clearer picture of noncopyable generics
```
I think other parts of the code won't expect to see an abstract conformance when the conforming type is not a type parameter or archetype. Perhaps the swiftinterface printing code should be changed instead? I think it already knows to skip Sendable conformances in some cases IIRC.

kavon replied:
i will have to address that later. i'm not sure yet if we want concrete conformances at all. it's likely that Copyable will get revised into a LayoutConstraint.

slavapestov commented on lib/AST/Module.cpp:
```diff
@@ -1646,6 +1646,11 @@ LookupConformanceInModuleRequest::evaluate(
 
   auto nominal = type->getAnyNominal();
 
+  // If it's not a nominal, then it's safe to assume its copyable.
```
Tuples should probably not be copyable unless all of their fields are copyable; is that already handled?

kavon replied:
yes tuples and all of the other kinds of types checked-for just before this are aware of Copyable. I didn't really know how to handle pack types / `repeat while` so I just stuck this here to catch "types not otherwise handled". I can clarify that in the comment.

kavon replied:
actually i'll just put specific handling of `PackExpansionType` for Copyable since that's the only reason I put this catch-all here.

slavapestov replied:
Conformance lookup with a PackExpansionType should not actually occur; it would be fine to put an assertion guarding against that.

As for a PackType, it will conform to Copyable if every element conforms to Copyable. The existing pack conformance logic should handle that because it recursively looks up the conformance on each element of the pack.

hborla commented on test/type/pack_expansion.swift:
```diff
@@ -74,3 +74,9 @@ func invalidPackRef(_: each Int) {}
+// coverage to ensure a 'repeat each' type is considered Copyable
+func golden<Z>(_ z: Z) {}
+func hour<each T>(_ t: repeat each T)  {
+  golden(t)
```
Yeah, we still need to diagnose value pack references appearing outside of pack expansions. You can write this as

```swift
repeat golden(each t)
```

kavon replied:
Ah. Had no idea. Thanks!

kavon commented:
@swift-ci please test and merge

slavapestov commented on test/Constraints/members.swift:
```diff
@@ -737,7 +737,7 @@ func rdar66891544() {
   func foo<U>(_: U, defaultU: U? = nil) {}
 
-  foo(.bar) // expected-error {{cannot infer contextual base in reference to member 'bar'}}
+  foo(.bar) // expected-error {{type '_Copyable' has no member 'bar'}}
```
Perhaps we should more generally do this for all marker protocols; a diagnostic like `type 'Sendable' has no member 'bar'" would be somewhat unexpected too for the same reason.

xedin replied:
Yeah, implicit conformances to marker protocols should not propagate (I am not sure about Sendable, haven’t reviewed any PRs for it).

kavon replied:
sure, I could make it general to exclude all marker protocols.... but you're allowed to add members to a marker protocol via extension, though you can't add static members:

```
extension Sendable {
  var member: Int { return 100 }
  typealias buddy = Int
}

func take<T: Sendable>(_ t: T) {}

func pass() {
  take(.member)    // error: instance member 'member' cannot be used on type 'Sendable'
  take(.nonmember) // error: type 'Sendable' has no member 'nonmember'
  take(.buddy)     // OK
}
```

so if I exclude marker protocols, all three calls turn into errors with "cannot infer contextual base in reference to member ...".

Since _Copyable is marked unavailable, you can't extend it, so nobody can add members to it.

kavon replied:
Then again, I don't know why anyone would want to add members to Sendable so I guess not inferring it is fine. They need to make it explicit.

xedin replied:
Since it’s not possible to add static members it shouldn’t be included because leading-dot syntax in generic context only works when first member in the chain is static.

xedin replied:
Just to clarify - it shouldn’t be included if it’s not stated explicitly, your example should still infer Sendable.

kavon commented on test/Constraints/generics.swift:
```diff
@@ -933,7 +933,7 @@ func rdar79757320() {
 
   // FIXME: There has to be a way to propagate holes that makes it easy to suppress failures caused by missing members.
-  _ = Container(value: Value(42).formatted(.S(a: .a, b: .b(0)))) // expected-error {{type 'R_79757320' has no member 'S'}}
+  _ = Container(value: Value(42).formatted(.S(a: .a, b: .b(0)))) // expected-error {{reference to member 'S' cannot be resolved without a contextual type}}
```
thanks! that fixes it. I didn't know where this inference was happening and thought it wasn't simple to fix.

xedin replied:
No problem! Next time there are diagnostic regressions please ping me and I will help you out.

kavon committed:
add Copyable to a few `-parse-stdlib` tests

Since these tests may in the future have
a move-only type in it. If you're using
`-parse-stdlib -module-name Swift`
it's a good add Copyable, but not required,
at this time.

kavon commented:
@swift-ci please test and merge

kavon committed:
note a small diagnostic regression

the way these diagnostics were added seems to have been through score hacking
and I haven't been able to get the behavior back since introducing the
copyable type constraint for generics. I've logged restoring this as a TODO
to take care of sometime later when I have more time.

kavon commented:
@swift-ci please test macos platform

kavon commented:
@swift-ci please test linux platform

kavon commented:
turns out the `Swift-validation(linux-x86_64) :: SIL/parse_stdlib.sil` assertion failure is not caused by this PR. it's failing in general!

kavon commented:
@swift-ci please smoke test and merge
